txstratum/cli.py

```diff
@@ -51,110 +57,172 @@ def create_parser() -> ArgumentParser:
-        http_metrics = HttpPrometheusExporter(manager, pubsub, args.prometheus_port)
-        http_metrics.start()
-        logger.info('Prometheus metrics server running at 0.0.0.0:{}...'.format(args.prometheus_port))
-
-    if args.ban_addrs or args.ban_tx_ids:
-        tx_filters.append(FileFilter.load_from_files(args.ban_tx_ids, args.ban_addrs))
-
-    if args.toi_url or args.toi_apikey:
-        if not (args.toi_url and args.toi_apikey):
-            raise ValueError("Should pass both toi_url and toi_apikey")
-        toiclient = TOIAsyncClient(args.toi_url, args.toi_apikey)
-        tx_filters.append(TOIFilter(toiclient, block=args.toi_fail_block))
-
-    api_app = App(manager, max_tx_weight=args.max_tx_weight, max_timestamp_delta=args.max_timestamp_delta,
-                  tx_timeout=args.tx_timeout, fix_invalid_timestamp=args.fix_invalid_timestamp,
-                  only_standard_script=not args.allow_non_standard_script, tx_filters=tx_filters)
-    logger.info('API Configuration', max_tx_weight=api_app.max_tx_weight, tx_timeout=api_app.tx_timeout,
-                max_timestamp_delta=api_app.max_timestamp_delta, fix_invalid_timestamp=api_app.fix_invalid_timestamp,
-                only_standard_script=api_app.only_standard_script, tx_filters=tx_filters)
-
-    web_runner = web.AppRunner(api_app.app, logger=logger)
-    loop.run_until_complete(web_runner.setup())
-    site = web.TCPSite(web_runner, '0.0.0.0', args.api_port)
-    loop.run_until_complete(site.start())
-
-    try:
-        logger.info('Stratum Server running at 0.0.0.0:{}...'.format(args.stratum_port))
-        logger.info('TxMining API running at 0.0.0.0:{}...'.format(args.api_port))
-        if args.testnet:
+class RunService:
+    """This is the main class of the service. It starts everything up."""
+
+    manager: TxMiningManager
+    loop: AbstractEventLoop
+    tx_filters: List[TXFilter]
+
+    def __init__(self, args: Namespace) -> None:
+        """Initialize the service."""
+        self.args = args
+
+        self.loop = asyncio.get_event_loop()
+
+        self.pubsub = PubSubManager(self.loop)
+        self.backend = HathorClient(args.backend)
+        self.manager = TxMiningManager(
+            backend=self.backend,
+            pubsub=self.pubsub,
+            address=args.address,
+        )
+
+        self.configure_logging(args)
+
+    async def graceful_shutdown(self) -> None:
+        """Gracefully shutdown the service."""
+        logger.info('Gracefully shutting down...')
+
+        self.manager.refuse_new_jobs = True
+
+        while len(self.manager.tx_queue) > 0:
+            logger.info('Waiting for pending txs to finish...', txs_left=len(self.manager.tx_queue))
+            await asyncio.sleep(2)
+
+        self.manager.shutdown()
+
+    def sigterm_handler(self) -> None:
+        """Handle SIGTERM signal."""
+        logger.info('SIGTERM received.')
+
+        self.loop.create_task(self.graceful_shutdown())
```

Comment:
Any reason why sigterm calls `graceful_shutdown` and sigint `shutdown`? If that was a bug, maybe we could have a single method `handle_shutdown_signal`

Reply:
I guess you have a point.

I thought I had reasons to not run the full shutdown in Kubernetes, but after thinking about it again, I don't think there is any.